### DISCUSSION TITLE
Keeping connections open to TCP listener

### DISCUSSION THREAD

**Discussion by LauOVin** [nextgenhealthcare/connect#5238] [Q&A] [answered]
I have a Mirth channel that has TCP listener as source connector type, in server mode. We have nearly 200 medical devices that will connect to this channel approx. 10 times a day to send a HL7 message. 

We get a lot of error messages in the log, because it seems like the devices tries to keep the connection open and sends a keep alive signal every 5 minutes. The keep alive signal (or whatever it is) seems to not include end-bytes, which is expected from the MLLP protocol. This is the error:

```
[2022-06-09 03:55:56,404]  ERROR  (com.mirth.connect.connectors.tcp.TcpReceiver:746): Error receiving message (TCP Listener "Source" on channel c3109784-eab2-4439-8620-01eb3be00d1c).
com.mirth.connect.model.transmission.framemode.FrameStreamHandlerException: End of message bytes (<FS><CR>) not detected. Last bytes received: é)
	at com.mirth.connect.model.transmission.framemode.FrameStreamHandler.read(FrameStreamHandler.java:221)
	at com.mirth.connect.connectors.tcp.TcpReceiver$TcpReader.readBytes(TcpReceiver.java:813)
	at com.mirth.connect.connectors.tcp.TcpReceiver$TcpReader.call(TcpReceiver.java:620)
	at com.mirth.connect.connectors.tcp.TcpReceiver$TcpReader.call(TcpReceiver.java:511)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
```

Do any of you know how we can solve this? Should we change transmission mode to Basic TCP, or should we change Keep Connection Open to Yes? Other suggestions are very welcome :-)

Currently we don't know if we can reconfigure the devices to stop trying to keep the connection open.

Channel settings screenshot is attached.
We are using Mirth 3.11

![Channel settings](https://user-images.githubusercontent.com/107257827/173072449-7200b7ab-22c5-4fe5-9850-482eb951c5ff.png)
.

**Comment by siddharthsharma9258** [comment]
The very first google search reveals this

"Keep-alive packets MUST only be sent when no data or acknowledgement packets have been received for the connection within an interval"

So the only reason the other device is sending you a Keep Alive message, is probably because you are not sending it back any ACK message. You might want to change the Response drop down and make it as soon as you receive.

**Comment by LauOVin** [comment]
Hi @siddharthsharma9258 - thanks for your answer.

After the device has sent a real HL7 message to the channel, the channel sends a auto-generated response after the source transformer, and then the channel should close the connection. I don't know why the device would send keep-alive packets afterwards. But maybe this is not something that can be fixed in Mirth, but should be fixed in the device configurations :-)

**Reply by siddharthsharma9258** [reply]
Is it only one device that's doing this?

**Comment by LauOVin** [comment, marked as answer]
Error messages was causes by keep-alive messages. Spoke with the technical support from the vendor. Keep-alive messages was caused by the device trying to synchronize time. By enabling NTP in the device configuration, and setting a server hostname, the keep-alive messages to Mirth stopped.
